### PR TITLE
Stabilize impl Trait

### PR DESCRIPTION
**Pull request by cramertj**
Blocked on:

- [x] https://github.com/rust-lang/rust/pull/49041 and
- [x] completion of FCP in https://github.com/rust-lang/rust/issues/34511#issuecomment-373207183 (3 days from now).

I have not yet done any docs work for this-- I probably won't get to it until this weekend (might be a project for the flight to the all-hands).

### CODE REVIEW

**Comment by cramertj** [comment]
Also-- miri and clippy both used `conservative_impl_trait` internally. I removed those locally, but I'm not sure of the right way to update those submodules atomically w/ the feature removal.

**Comment by nikomatsakis** [comment]
With respect to docs, @steveklabnik plans to file an RFC changing process there somewhat in any case.

File: src/libsyntax/feature_gate.rs [hunk 1]
```diff
@@ -556,6 +550,10 @@ declare_features! (
     (accepted, inclusive_range_syntax, "1.26.0", Some(28237), None),
     // allow `..=` in patterns (RFC 1192)
     (accepted, dotdoteq_in_patterns, "1.26.0", Some(28237), None),
+    // Allows `impl Trait` in function arguments.
+    (accepted, universal_impl_trait, "1.26.0", Some(34511), None),
+    // Allows `impl Trait` in function return types.
+    (accepted, conservative_impl_trait, "1.26.0", Some(34511), None),
```
Comment (nikomatsakis) [inline src/libsyntax/feature_gate.rs]: 🚀

**Comment by nikomatsakis** [comment]
@bors r+

The FCP is not yet complete, but it's hard to imagine what would cause us to reverse course here, and in any case we can always backport a patch to beta reversing course or fixing things up.

**Comment by nikomatsakis** [comment]
(cc @Manishearth and @oli-obk for clippy/miri respectively)

**Comment by Manishearth** [comment]
@cramertj we'll just update, thanks.

The best way to do it is to throw us a PR. even better is to update rustc to point to the commit in the PR so we don't need to come back and do that. But you don't have to.

**Comment by kennytm** [comment]
Stabilize ALL The Things!

**Comment by mark-i-m** [comment]
1.26 is going to be a _really_ exciting release. We already have i128 and `..=` in there...

**Comment by cramertj** [comment]
@mark-i-m And `!`-type, and probably `dyn` and `match_default_bindings`, slice patterns...

**Comment by mark-i-m** [comment]
Woah what really!? :tada:

**Comment by kennytm** [comment]
[Confirmed stabilized](https://github.com/rust-lang/rust/issues?utf8=%E2%9C%93&q=is%3Amerged+label%3Arelnotes+merged%3A%3E%3D2018-02-15): 

* `Option<&mut T>::cloned()`
* `LocalKey::try_with` (for thread-local storage)
* **slice pattern (`match x { &[y, z] => w, ... }`)**
* `[T]::rotate_left` and `[T]::rotate_right`
* The Book 2nd edition
* `--remap-path-prefix`
* `<*T>::add`, `<*T>::copy_nonoverlapping`, etc
* `Box::leak`
* inclusive range (`..=`)
* never type (`!`)
* `FusedIterator`

[Scheduled to be stabilized](https://github.com/rust-lang/rust/issues?utf8=%E2%9C%93&q=is%3Aopen+label%3Arelnotes+label%3AS-waiting-on-bors):

* i128
* impl Trait

I'm not quite sure if `dyn` and `match_default_bindings` can make it, as 1.26 is going to enter beta on March 27th (Tuesday). We could keep 1.27 exciting??? 😄

**Comment by mark-i-m** [comment]
Lol @nrc is this boring enough for you :)

File: src/libsyntax/feature_gate.rs (hunk 1 above)
Comment (0zero0zero) [reply src/libsyntax/feature_gate.rs]: wowoow... finally.

**Comment by SimonSapin** [comment]
Reviewers! Help me add to that list ;)

* `Copy` and `Clone` for closures https://github.com/rust-lang/rust/pull/49299
* `TryFrom` and `TryInto` https://github.com/rust-lang/rust/pull/49305

**Comment by kennytm** [comment]
I've also got the following [errors](https://travis-ci.org/rust-lang/rust/builds/357803153) in the rollup, complaining about `stable_features`. Please update these tests after rebasing.

```
[00:40:16] failures:
[00:40:16]     [ui] ui/impl-trait/region-escape-via-bound-contravariant-closure.rs
[00:40:16]     [ui] ui/impl-trait/region-escape-via-bound-contravariant.rs
```

**Comment by SimonSapin** [comment]
Should we sort entries in that file by feature name rather, to try an minimize merge conflicts?

**Comment by SimonSapin** [comment]
```
[01:18:41] failures:
[01:18:41] 
[01:18:41] ---- [run-pass] run-pass/needless_lifetimes_impl_trait.rs stdout ----
[01:18:41]      
[01:18:41] error: compilation failed!
[01:18:41] status: exit code: 101
[01:18:41] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin/clippy-driver" "tests/run-pass/needless_lifetimes_impl_trait.rs" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/re
lease/build/clippy-49a9977e3d89da30/out/test_build_base" "--target=x86_64-unknown-linux-gnu" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/build/clippy-49a9977e3d89da
30/out/test_build_base/needless_lifetimes_impl_trait.stage-id" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps" "-Dwarnings" "-L" "/check
out/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/build/clippy-49a9977e3d89da30/out/test_build_base/needless_lifetimes_impl_trait.stage-id.aux"
[01:18:41] stdout:
[01:18:41] ------------------------------------------
[01:18:41] 
[01:18:41] ------------------------------------------
[01:18:41] stderr:
[01:18:41] ------------------------------------------
[01:18:41] error: this feature has been stable since 1.26.0. Attribute no longer needed
[01:18:41]  --> tests/run-pass/needless_lifetimes_impl_trait.rs:3:12
[01:18:41]   |
[01:18:41] 3 | #![feature(conservative_impl_trait)]
[01:18:41]   |            ^^^^^^^^^^^^^^^^^^^^^^^
[01:18:41]   |
[01:18:41]   = note: `-D stable-features` implied by `-D warnings`
[01:18:41] 
[01:18:41] error: aborting due to previous error
[01:18:41] 
[01:18:41] 
[01:18:41] ------------------------------------------
[01:18:41] 
[01:18:41] thread '[run-pass] run-pass/needless_lifetimes_impl_trait.rs' panicked at 'explicit panic', /cargo/registry/src/github.com-1ecc6299db9ec823/compiletest_rs-0.3.8/src/runtest.rs:2539:9
[01:18:41] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:18:41] 
[01:18:41] 
[01:18:41] failures:
[01:18:41]     [run-pass] run-pass/needless_lifetimes_impl_trait.rs
[01:18:41] 
[01:18:41] test result: FAILED. 11 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
[01:18:41] 
[01:18:41] test compile_test ... FAILED
[01:18:41] 
[01:18:41] failures:
[01:18:41] 
[01:18:41] ---- compile_test stdout ----
[01:18:41]      thread 'compile_test' panicked at 'Some tests failed', /cargo/registry/src/github.com-1ecc6299db9ec823/compiletest_rs-0.3.8/src/lib.rs:89:22
[01:18:41] 
[01:18:41] 
[01:18:41] failures:
[01:18:41]     compile_test
[01:18:41] 
[01:18:41] test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
[01:18:41] 
[01:18:41] error: test failed, to rerun pass '--test compile-test'
[01:18:41] 
[01:18:41] 
[01:18:41] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "--target" "x86_64-unknown-linux-gnu" "-j" "4" "--release" "--locked" "--color" "always" "--manifest-path" "/checkout/src/tools/clippy/Cargo.toml"

**Comment by cramertj** [comment]
Hm... that test is coming from clippy, which is in a submodule. Do I need to send and merge a clippy patch and then add a submodule update to this PR, or is there a way to temporarily ignore the test? (without modifying the test itself that is, b.c. that would have the same issue)

**Comment by oli-obk** [comment]
Öhhhhh. Clippy should not be failing any builds. @kennytm any idea what's happening?

**Comment by kennytm** [comment]
@oli-obk we disallow regressing the tools in the last week of the 6-week cycle.

**Comment by oli-obk** [comment]
Oh. Neat. I added an allow stable features attr in https://github.com/rust-lang-nursery/rust-clippy/pull/2575 you can point the submodule to clippy master now

**Comment by cramertj** [comment]
Oh awesome, thanks-- I'll do that now.

**Comment by cramertj** [comment]
Opened a PR doing the same for miri [here](https://github.com/solson/miri/pull/365)

**Comment by oli-obk** [comment]
I don't think  that's necessary, miri was broken before and still is, so no regression

**Comment by oli-obk** [comment]
I think I even removed miri from bors until we get it fixed

**Comment by cramertj** [comment]
@bors r-

need to update lockfile first :)

**Pull request merged**